File: src/main/java/io/scriptor/impl/device/CLINT.java
package io.scriptor.impl.device;

import io.scriptor.fdt.BuilderContext;
import io.scriptor.fdt.NodeBuilder;
import io.scriptor.machine.Device;
import io.scriptor.machine.IODevice;
import io.scriptor.machine.Machine;
import io.scriptor.util.Log;
import org.jetbrains.annotations.NotNull;

import java.io.PrintStream;
import java.util.Arrays;

public final class CLINT implements IODevice {

    private static final int MSIP_BASE = 0x0000;
    private static final int MSIP_STRIDE = 4;
    private static final int MTIMECMP_BASE = 0x4000;
    private static final int MTIMECMP_STRIDE = 8;
    private static final int CONTEXT_BASE = 0xB000;
    private static final int MTIME_OFFSET = 0xBFF8;

    private final Machine machine;
    private final long begin;
    private final long end;
    private final int hartCount;

    private long mtime;

    private final long[] mtimecmp;

    /**
     * machine-level external interrupt pending bit
     */
    private final boolean[] meip;
    /**
     * machine-level software interrupt pending bit
     */
    private final boolean[] msip;

    public CLINT(final @NotNull Machine machine, final long begin, final int hartCount) {
        this.machine = machine;
        this.begin = begin;
        this.end = begin + 0x10000L;
        this.hartCount = hartCount;

        this.mtime = 0L;
        this.mtimecmp = new long[hartCount];

        this.meip = new boolean[hartCount];
        this.msip = new boolean[hartCount];
    }

    @Override
    public @NotNull Machine machine() {
        return machine;
    }

    @Override
    public void dump(final @NotNull PrintStream out) {
        out.printf("clint: mtime=%x%n", mtime);
        for (int id = 0; id < hartCount; ++id) {
            out.printf("#%-2d | mtimecmp=%x meip=%b msip=%b%n", id, mtimecmp[id], meip[id], msip[id]);
        }
    }

    @Override
    public void reset() {
        mtime = 0L;
        Arrays.fill(mtimecmp, ~0L);
        Arrays.fill(meip, false);
        Arrays.fill(msip, false);
    }

    @Override
    public void step() {
        ++mtime;
    }

    @Override
    public void build(final @NotNull BuilderContext<Device> context, final @NotNull NodeBuilder builder) {
        final var phandle = context.push(this);

        final var ie = new int[4 * hartCount];
        for (int i = 0; i < hartCount; ++i) {
            final var cpu = context.get(machine.hart(i));
            ie[i * 4] = cpu;
            ie[i * 4 + 1] = 0x03;
            ie[i * 4 + 2] = cpu;
            ie[i * 4 + 3] = 0x07;
        }

        builder.name("clint@%x".formatted(begin))
               .prop(pb -> pb.name("phandle").data(phandle))
               .prop(pb -> pb.name("compatible").data("riscv,clint0"))
               .prop(pb -> pb.name("reg").data(begin, end - begin))
               .prop(pb -> pb.name("interrupts-extended").data(ie));
    }

    @Override
    public long begin() {
        return begin;
    }

    @Override
    public long end() {
        return end;
    }

    @Override
    public long read(final int offset, final int size) {
        if (offset >= MSIP_BASE && offset < MTIMECMP_BASE && size == 4) {
            final var hart = (offset - MSIP_BASE) / MSIP_STRIDE;
            if (hart >= hartCount) {
                return 0L;
            }
            return msip[hart] ? 1L : 0L;
        }

        if (offset >= MTIMECMP_BASE && offset < CONTEXT_BASE && size == 8) {
            final var hart = (offset - MTIMECMP_BASE) / MTIMECMP_STRIDE;
            if (hart >= hartCount) {
                return 0L;
            }
            return mtimecmp[hart];
        }

        if (offset == MTIME_OFFSET && size == 8) {
            return mtime;
        }

        Log.error("invalid clint read offset=%x, size=%d", offset, size);
        return 0L;
    }

    @Override
    public void write(final int offset, final int size, final long value) {
        if (offset >= MSIP_BASE && offset < MTIMECMP_BASE && size == 4) {
            final var hart = (offset - MSIP_BASE) / MSIP_STRIDE;
            if (hart >= hartCount) {
                return;
            }
            msip[hart] = value != 0L;
            return;
        }

        if (offset >= MTIMECMP_BASE && offset < CONTEXT_BASE && size == 8) {
            final var hart = (offset - MTIMECMP_BASE) / MTIMECMP_STRIDE;
            if (hart >= hartCount) {
                return;
            }
            mtimecmp[hart] = value;
            return;
        }

        Log.error("invalid clint write offset=%x, size=%d, value=%x", offset, size, value);
    }

    @Override
    public @NotNull String toString() {
        return "clint@%x".formatted(begin);
    }

    public long mtime() {
        return mtime;
    }

    public long mtimecmp(final int id) {
        return mtimecmp[id];
    }

    public void mtimecmp(final int id, final long value) {
        mtimecmp[id] = value;
    }

    public boolean meip(final int id) {
        return meip[id];
    }

    public boolean mtip(final int id) {
        return Long.compareUnsigned(mtime, mtimecmp[id]) >= 0;
    }

    public boolean msip(final int id) {
        return msip[id];
    }

    public void msip(final int id, final boolean value) {
        msip[id] = value;
    }
}

File: src/main/java/io/scriptor/impl/device/CLINT.kt
package io.scriptor.impl.device

import io.scriptor.fdt.BuilderContext
import io.scriptor.fdt.NodeBuilder
import io.scriptor.machine.Device
import io.scriptor.machine.IODevice
import io.scriptor.machine.Machine
import io.scriptor.util.Log.error
import io.scriptor.util.Log.format
import java.io.PrintStream

class CLINT : IODevice {

    override val machine: Machine

    override val begin: ULong
    override val end: ULong

    private var mtime: ULong = 0UL

    @OptIn(ExperimentalUnsignedTypes::class)
    private val mtimecmp: ULongArray

    /**
     * machine-level external interrupt pending bit
     */
    private val meip: BooleanArray

    /**
     * machine-level software interrupt pending bit
     */
    private val msip: BooleanArray

    @OptIn(ExperimentalUnsignedTypes::class)
    constructor(machine: Machine, begin: ULong) {
        this.machine = machine
        this.begin = begin
        this.end = begin + 0x10000UL

        this.mtimecmp = ULongArray(machine.harts.size)
        this.meip = BooleanArray(machine.harts.size)
        this.msip = BooleanArray(machine.harts.size)
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun dump(out: PrintStream) {
        out.println(format("clint: mtime=%x", mtime))
        for (id in 0 until machine.harts.size) {
            out.println(
                format(
                    "#%-2d | mtimecmp=%x meip=%b msip=%b",
                    id,
                    mtimecmp[id],
                    meip[id],
                    msip[id],
                ),
            )
        }
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun reset() {
        mtime = 0UL
        mtimecmp.fill(0UL.inv())
        meip.fill(false)
        msip.fill(false)
    }

    override fun step() {
        ++mtime
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun build(context: BuilderContext<Device>, builder: NodeBuilder) {
        val phandle = context.get(this)

        val ie = UIntArray(4 * machine.harts.size)
        for (id in 0 until machine.harts.size) {
            val cpu = context.get(machine.harts[id])
            ie[id * 4] = cpu
            ie[id * 4 + 1] = 0x03U
            ie[id * 4 + 2] = cpu
            ie[id * 4 + 3] = 0x07U
        }

        builder
            .name(format("clint@%x", begin))
            .prop { it.name("phandle").data(phandle) }
            .prop { it.name("compatible").data("riscv,clint0") }
            .prop { it.name("reg").data(begin, end - begin) }
            .prop { it.name("interrupts-extended").data(*ie) }
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun read(offset: UInt, size: UInt): ULong {
        if (offset in MSIP_BASE until MTIMECMP_BASE && size == 4U) {
            val hart = ((offset - MSIP_BASE) / MSIP_STRIDE).toInt()
            if (hart >= machine.harts.size) return 0U
            return if (msip[hart]) 1U else 0U
        }

        if (offset in MTIMECMP_BASE until CONTEXT_BASE && size == 8U) {
            val hart = ((offset - MTIMECMP_BASE) / MTIMECMP_STRIDE).toInt()
            if (hart >= machine.harts.size) return 0U
            return mtimecmp[hart]
        }

        if (offset == MTIME_OFFSET && size == 8U) {
            return mtime
        }

        error("invalid clint read offset=%x, size=%d", offset, size)
        return 0U
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    override fun write(offset: UInt, size: UInt, value: ULong) {
        if (offset in MSIP_BASE until MTIMECMP_BASE && size == 4U) {
            val hart = ((offset - MSIP_BASE) / MSIP_STRIDE).toInt()
            if (hart >= machine.harts.size) return
            msip[hart] = (value != 0UL)
            return
        }

        if (offset in MTIMECMP_BASE until CONTEXT_BASE && size == 8U) {
            val hart = ((offset - MTIMECMP_BASE) / MTIMECMP_STRIDE).toInt()
            if (hart >= machine.harts.size) return
            mtimecmp[hart] = value
            return
        }

        error("invalid clint write offset=%x, size=%d, value=%x", offset, size, value)
    }

    override fun toString(): String {
        return format("clint@%x", begin)
    }

    fun mtime(): ULong {
        return mtime
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    fun mtimecmp(id: Int): ULong {
        return mtimecmp[id]
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    fun mtimecmp(id: Int, value: ULong) {
        mtimecmp[id] = value
    }

    fun meip(id: Int): Boolean {
        return meip[id]
    }

    @OptIn(ExperimentalUnsignedTypes::class)
    fun mtip(id: Int): Boolean {
        return mtime >= mtimecmp[id]
    }

    fun msip(id: Int): Boolean {
        return msip[id]
    }

    fun msip(id: Int, value: Boolean) {
        msip[id] = value
    }

    companion object {
        private const val MSIP_BASE = 0x0000U
        private const val MSIP_STRIDE = 4U
        private const val MTIMECMP_BASE = 0x4000U
        private const val MTIMECMP_STRIDE = 8U
        private const val CONTEXT_BASE = 0xB000U
        private const val MTIME_OFFSET = 0xBFF8U
    }
}
